test/Transactions/Orleans.Transactions.Tests/Grains/TransactionAttributionGrain.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Orleans.Transactions.Tests.Grains
+{
+    public class NoAttributionGrain : Grain, INoAttributionGrain
+    {
+        public Task<Dictionary<int, List<string>>> GetNestedTransactionIds(int tier, Dictionary<int, List<ITransactionAttributionGrain>> tiers)
+        {
+            return AttributionGrain.GetNestedTransactionIds(tier, tiers);
+        }
+    }
+
+    public class NotSupportedAttributionGrain : Grain, INotSupportedAttributionGrain
+    {
+        public Task<Dictionary<int, List<string>>> GetNestedTransactionIds(int tier, Dictionary<int, List<ITransactionAttributionGrain>> tiers)
+        {
+            return AttributionGrain.GetNestedTransactionIds(tier, tiers);
+        }
+    }
+
+    public class RequiredAttributionGrain : Grain, IRequiredAttributionGrain
+    {
+        public Task<Dictionary<int, List<string>>> GetNestedTransactionIds(int tier, Dictionary<int, List<ITransactionAttributionGrain>> tiers)
+        {
+            return AttributionGrain.GetNestedTransactionIds(tier, tiers);
+        }
+    }
+
+    public class RequiresNewAttributionGrain : Grain, IRequiresNewAttributionGrain
+    {
+        public Task<Dictionary<int, List<string>>> GetNestedTransactionIds(int tier, Dictionary<int, List<ITransactionAttributionGrain>> tiers)
+        {
+            return AttributionGrain.GetNestedTransactionIds(tier, tiers);
+        }
+    }
+
+    internal static class AttributionGrain
+    {
+        static public async Task<Dictionary<int, List<string>>> GetNestedTransactionIds(int tier, Dictionary<int, List<ITransactionAttributionGrain>> tiers)
+        {
+            ITransactionInfo ti = TransactionContext.GetTransactionInfo();
+            Dictionary<int, List<string>> results = new Dictionary<int, List<string>>();
+            results[tier] = new List<string>(new[] { ti?.Id });
+
+            if (tiers.Count == 0)
+            {
+                return results;
+            }
+
+            KeyValuePair<int, List<ITransactionAttributionGrain>> nextTier = tiers.FirstOrDefault();
```

Comment:
Should this also be ordered?

Reply:
I will move to using flat array, will make it more clear.  Started that way and moved to dictionary for ease of coding, but it adds confusion.